### PR TITLE
add tests for validating that there are consistent methods within the assertions

### PR DESCRIPTION
**Pull request by filiphr**
This is are proposal tests for validating that the assertions have the correct methods.
related to #671

@joel-costigliola I would appreciate if you can have a look at it and let me know what you think. There are some asserts which are in the static assertions but not in the soft.

The ```XXXNavigableXXXAssert```(s) and some others are missing. 

I can start on "fixing" the tests so they do not fail, but I would need some guidance.

### CODE REVIEW

**Comment by joel-costigliola** [comment]
Nice work ! I like the approach, it showed an inconsistency between `Assertions` and `BDDAssertions` (the `Map` one that I'm fixing right now).

Getting the test on soft assertions working is gonna be more challenging, the `XXXNavigableXXXAssert` are not yet in soft assertions, I tried to add them but failed due to generic problems, I might give another try later (aah generic hell !).

I suggest to exclude `XXXNavigableXXXAssert` from the analysis and to define another method comparator that would ignore return types until we change soft assertions to return the same type as standard assertions.

When checking `Assertions` vs `AbstractStandardSoftAssertions`, no need to ignore the method names as they should identical.

How about splitting this PR in two, first part for checking `Assertions` and `BDDAssertions` and create another PR for soft assertions (would also include testing `AbstractBDDSoftAssertions`) ?

File: src/test/java/org/assertj/core/api/AssertionsTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  public void should_have_the_same_methods_as_in_bdd_assertions() {
+    Class<Assertions> classA = Assertions.class;
+    String methodA = "assertThat";
+    Class<BDDAssertions> classB = BDDAssertions.class;
+    String methodB = "then";
+    List<Method> assertThatMethods = findMethodsWithName(classA, methodA);
+    List<Method> thenMethods = findMethodsWithName(classB, methodB);
+
+    Comparator<Method> methodComparator = ignoringDeclaringClassAndMethodName();
+    Assertions.assertThat(assertThatMethods).usingElementComparator(methodComparator)
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/AssertionsTest.java]: I would make `findMethodsWithName` to return `Method[]`, that would avoid the array conversion in the tests.

Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/AssertionsTest.java]: I would inverse `assertThatMethods` and `thenMethods` in the assertion, we are testing `BDDAssertions` methods not `Assertions` ones.
And also, let's static import `assertThat`, so much nicer to the eye ;-)

File: src/test/java/org/assertj/core/api/AssertionsTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+public class AssertionsTest extends BaseAssertionsTest {
+
+  @Test
+  public void should_have_the_same_methods_as_in_bdd_assertions() {
+    Class<Assertions> classA = Assertions.class;
+    String methodA = "assertThat";
+    Class<BDDAssertions> classB = BDDAssertions.class;
+    String methodB = "then";
+    List<Method> assertThatMethods = findMethodsWithName(classA, methodA);
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/AssertionsTest.java]: inline local variables `classA`, `classB`, `methodA` and `methodB`

File: src/test/java/org/assertj/core/api/BaseAssertionsTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        matchingMethods.add(method);
+      }
+    }
+    return matchingMethods;
+  }
+
+  protected static Comparator<Method> ignoringDeclaringClassAndMethodName() {
+    return new Comparator<Method>() {
+      @Override
+      public int compare(Method o1, Method o2) {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/BaseAssertionsTest.java]: rename `o1` to `method1` and `o2` to `method2` for better code readability.

**Comment by filiphr** [comment]
Regarding the inherited methods, don't we also need to have ```Java6AbstractBDDSoftAssertions```. I will modify the ```findMethodWithName``` to search for declared methods up the inheritance tree.

I will add the possibility to ```findMethodWithName``` to ignore methods when the return type is instance of something passed. Will also add the second method comparator ignoring the return type.

I have no problem splitting the PR. I will remove the soft assertions tests from this one and do a force push.

File: src/test/java/org/assertj/core/api/AssertionsTest.java (hunk 1 above)
Comment (filiphr) [inline src/test/java/org/assertj/core/api/AssertionsTest.java]: There is an inverse test in the ```BDDAssertionsTest```.  I am not sure if we need to test in both places though.  

Will do the static import, I am not sure how it slipped :).

**Comment by joel-costigliola** [comment]
:+1: 

Indeed, I have reopened https://github.com/joel-costigliola/assertj-core/issues/672 for `Java6AbstractBDDSoftAssertions`.

**Comment by filiphr** [comment]
So do we make ```findMethodWithName``` go up the inheritance tree or we should test each level separately?

**Comment by joel-costigliola** [comment]
I prefer going up the inheritance tree as it ends up testing the concrete classes users use.

**Comment by filiphr** [comment]
Did that. I think that this covers the ```Assertions``` vs ```BDDAssertions``` method validation.

Once this is merged, I will make a pull request with the ```SoftAssertions``` vs the ```BDDSoftAssertions``` methods validations test.

**Comment by joel-costigliola** [comment]
Merged, I just haven't included `BDDSoftAssertionsTest` and `SoftAssertionsTest` as they should be part of the next PR.

**Comment by filiphr** [comment]
@joel-costigliola did you maybe forgot to push the merge? I cannot see it.

**Comment by PascalSchumacher** [comment]
@filiphr joel just pushed to the 2.x branch: https://github.com/joel-costigliola/assertj-core/commit/c8dff5be837c439aea97d7acf1f01f0afe7812a3 he did not merge the commit to master yet

**Comment by filiphr** [comment]
@PascalSchumacher I didn't see that. Thanks for the information.

**Comment by filiphr** [comment]
@joel-costigliola Maybe we need to keep this open until we merge it into master. I didn't check the tests against master. 

I can do this tonight and give you an analysis.

**Comment by joel-costigliola** [comment]
Thanks but no need, I usually merge 2.x branch in master and fix/update whatever breaks.

**Pull request closed without merging**